### ISSUE TITLE
Transfer-Encoding Header For PUT request to AWS S3

### ISSUE THREAD

**Issue by MQZhangThu** [wkh237/react-native-fetch-blob#94]
I used pre-singed url to PUT image to AWS S3. The PUT request did not work  out, and the AWS S3 response is 501 Error, saying that "Transfer-Encoding: Chunked" Header is not implemented.

Anyone knows how to solve this problem? Thx.

**Comment by wkh237** [comment]
@MQZhangThu , thanks for reporting the issue :+1:  looks like it's related to [https://github.com/square/okhttp/issues/2190](https://github.com/square/okhttp/issues/2190), there's already a PR(#86) to solve this issue, but I don't have time to test and merge it, I'll update the status in this thread once it got merged.

**Comment by wkh237** [comment]
After some test, I found that when simply overrides `contentLength()` can disable the chunk encoding. However, the given content length should be correct otherwise it will lead to an `unexpected end of stream` error.

**Comment by MQZhangThu** [comment]
:+1:  so what is the recommend way to set the content length? One possible way is seting the "content-length" header as the byte number when call RNFetchBlob.fetch?  However, this may be a little weird, as in my opinion, the "content-length" should be set by RNFetchBlob lib itself, not by the user. :)

**Comment by wkh237** [comment]
Yeah, but I can't calculate correct content length from multipart body so far, I think it might takes some time.

**Comment by wkh237** [comment]
@MQZhangThu , I've published `0.9.2-beta.4` to npm, the requests in Android should be fixed length request now. I wish you could try install this version and help us verify if the problem has been resolved, Thank you ! 😃

**Comment by MQZhangThu** [comment]
@wkh237 It works now. Great.  But the `uploadProgress` seems not working well.

**Comment by wkh237** [comment]
@MQZhangThu , thank you for the information, I've fixed the bug in `0.9.2-beta.5`, please install this version to repair the function.

**Comment by adbl** [comment]
I'm getting 501:s when trying to `put` to S3 with pre-signed urls on iOS too...

**Comment by wkh237** [comment]
@adbl , IOS network module should not send chunked request, is there any way for me to reproduce this issue ?

**Comment by adbl** [comment]
hmm, pretty complicated if you're not already using signed AWS S3 urls.

I would be able to debug it with XCode if you can give me some instructions?

**Comment by wkh237** [comment]
@adbl , after some investigation I know where's the problem now. In our IOS native implementation, if you use an URI as the request body (use RNFetchBlob.wrap), we will create a file stream for it then pipe it to request stream for better memory efficiency [(refer to the implementation)](https://github.com/wkh237/react-native-fetch-blob/blob/master/src/ios/RNFetchBlobReqBuilder.m#L114). 

However, the API we use (NSURLSession) will automatically use `Chunked` encoding when the request body comes from a file stream. For example, consider the following code snippet

```js
let rnfbURI = RNFetchBlob.wrap(pathOfAFile)
RNFetchBlob.fetch('POST', `${TEST_SERVER_URL}/upload`, {}, rnfbURI)
```
which wil send a request with the following headers

<img width="520" alt="2016-08-31 10 11 32" src="https://cloud.githubusercontent.com/assets/5063785/18132125/072cc7c6-6fc8-11e6-8d7b-fb137a612560.png">

however, if we change the way we send request, use BASE64 encoded data

```js
RNFetchBlob .fetch('POST', `${TEST_SERVER_URL}/upload`, {
    'Content-Type' : 'application/octet-stream;BASE64'
}, base64)
```

it no longer send chunked request 

<img width="519" alt="2016-08-31 10 16 36" src="https://cloud.githubusercontent.com/assets/5063785/18132250/a4514fc2-6fc8-11e6-9506-19f8334051b3.png">

there's no such an option to manually change the way it send request at this moment, but I think it's great to add one so that it can be more flexible. 

If you have any idea about the concept please feel free to leave any comment, thank you 😄

**Comment by adbl** [comment]
Aha, is it possible to tell `NSURLSession` to not use `Chunked` encoding when it is sending a stream? If so, there could be option to `config` which forces to not use chunked encoding?

For my purposes, bas64 is no-go.

**Comment by wkh237** [comment]
From my understanding, if not using `Chunked` encoding, the request data must be sent with fix content length, so that server knows when to close the stream. I've found [this document](http://docs.aws.amazon.com/AmazonS3/latest/dev/UsingRESTAPImpUpload.html) it suggets use multipart upload to upload large file, I think this is very common. Not sure if this is the right one?

In fact there's an fs API `fs.slice` which already included in 0.9.4 package but not addressed in document. We've use this API along with FireBase SDK and looks like it's working properly. You may refer to [the test case](https://github.com/wkh237/react-native-fetch-blob/blob/master/test/test-fs.js#L412). By use it, you can split the file into small chunks then upload them separately.

**Comment by adbl** [comment]
Aha cool, that's great when uploading big files. Our particular use case they are only around 400kb. Would it be feasible to implement this force thing (I guess that means it have to read the file into memory to calculate length. But for small files that might be ok?

**Comment by adbl** [comment]
As long as it's done in native I mean.

**Comment by wkh237** [comment]
You may take a look at [this issue](https://github.com/wkh237/rn-firebase-storage-upload-sample/issues/2). The way that FireBase SDK deal with different size files, it send the file in single request if the size is smaller than 256kb otherwise it splits the file into 256kb chunks then upload them separately. I think this is a good reference 😏 

On the other hand, I will add an additional option for developers to switch on/off `Chunked` transfer encoding 👍

**Comment by adbl** [comment]
Nice!

**Comment by wkh237** [comment]
The fix has been released with `0.9.4` so I'm going to close this issue. Please feel free reopen the issue if there's any problem, thank you 😄
